Add a Macros chapter (#137)
apple/swift-book: TSPL.docc/LanguageGuide/Macros.md

that expands the macro by generating Swift code,
and its declaration 
as well as the macro's name, what parameters it takes,
where it can be used,
and what kind of code it generates.
<!-- XXX typically (always?) in different libraries -->

You introduce a macro declaration with the `macro` keyword.
For example,
here's part of the declaration for
the `@OptionSet` macro used in the previous example:

```swift
public macro OptionSet<RawType>() =
        #externalMacro(module: "SwiftMacros", type: "OptionSetMacro")
```

The first line
specifies the macro's name, `OptionSet`,
and the arguments that takes, in this case none.
The second line
uses the `#externalMacro(module:type:)` macro from the standard library
to tell Swift where the macro's implementation is.
In this case,
the `SwiftMacros` module
contains a type named `OptionSetMacro`,
which implements the `@OptionSet` macro.

<!-- XXX TR:
Is there any other syntax besides #externalMacro
that's common enough to call out in the guide,
versus just being covered in the reference?
-->

> Note:
> Macros are always declared as `public`.
> Because the code that declares a macro
> is in a different module from code that uses that macro,

> > there isn't anywhere you could apply a non-public macro.

Author: "Non-public" includes `fileprivate`, `internal`, and `private` — every access level except `public` will make your macro not work.

```
@freestanding(expression)
public macro line<T: ExpressibleByIntegerLiteral>() -> T =
        /* ... location of the macro implementation... */
```

A *macro role* indicates where the code that the macro generate goes.
Every macro has one or more roles,
which you write as part of the attributes
at the beginning of the macro declaration.

For example,
the `#line` macro above has the `expression` role,
and the `@OptionSet` macro above has two roles --- `conformance` and `member`.
An expression macro produces a value or performs an action.
A conformance macro adds one or more protocol conformances.
The `@OptionSet` macro
extends the type that you apply the macro to,
to add conformance to the `OptionSet` protocol.
A member macro adds new members to the type you apply the macro to.
In this case,
the `@OptionSet` macro adds an `init(rawValue:)` initializer
that's required by the `OptionSet` protocol,
as well as some additional members.

In addition to the macro's role,
a macro's declaration provides information about
the names of the symbols that the macro generates.
When a macro declaration provides a list of names,
it's guaranteed to produce only declarations that use those names,

> which helps you understand and debug the generated code.

Reviewer: NICE!

The macro adds declarations for symbols named
`RawValue`, `rawValue`, and `init` ---
because those names are known ahead of time,
the macro declaration lists them explicitly.

The macro declaration also includes `arbitrary` after the list of names,
allowing the macro to generate declarations
whose names aren't known until you use the macro.
In this case,
for each case in the private `Options` enumeration,
`@OptionSet` generates a corresponding declaration with the same name.

For more information,
including a full list of macro roles,
see <doc:Attributes#attached> and <doc:Attributes#freestanding>
in <doc:Attributes>

## Macro Expansion

As part of building Swift code that uses macros,
the compiler and the macro's implementation
pass that code back and forth to expand the macros.

Macros are expanded as follows:

1. The compiler reads the code,
   creating an in-memory representation of the syntax.

1. The compiler sends part of the in-memory representation
   to the macro implementation,
   which expands the macro.

1. The compiler replaces the macro call with its expanded form.

1. The compiler continues with compilation,
   using the expanded source code.

> ![](macro-expansion-full)

Reviewer: Please add alt-text.

To go through the specific steps,

> consider the following code:

Reviewer: ```suggestion
consider the following:
```

```
let magicNumber = #fourCharacterCode("ABCD")
```

The `#fourCharacterCode` macro takes a string that's four characters long

> and returns a unsigned 32-bin integer

Reviewer: Assume that should be "32-bit"?

that corresponds to the ASCII values in the string joined together.

> Some file formats use a integers like this to identify data,

Reviewer: ```suggestion
Some file formats use integers like this to identify data,
```

> Some file formats use a integers like this to identify data,

Reviewer: ```suggestion
Some file formats use integers like this to identify data
```

because they're compact but still readable in a debugger.
The <doc:Macros#Implementing-Macros> section below
shows how to implement this macro.

To expand the macros in the code above,
the compiler reads the Swift file
and creates an in-memory representation of that code

> known an as *abstract syntax tree* or AST.

Reviewer: ```suggestion
known an as *abstract syntax tree*, or AST.
```

The AST makes the code's meaning and structure explicit,
which makes it easier to write code that interacts with that structure ---
like a compiler or a macro implementation.
Here's a representation of the AST for the code above,

> slightly simplified by omitting some extra detail.

Reviewer: ```suggestion
slightly simplified by omitting some extra detail:
```

[XXX placeholder art]

- Constant declaration
    - Identifier: `magicNumber`
    - Value
        - Macro
            - Identifier: `fourCharacterCode`
            - Arguments
                - String literal: `"ABCD"`

<!-- XXX
walk through the AST
don't assume people know how to read this kind of diagram
-->

> The compiler also ensures that the input code is valid Swift:

Reviewer: ```suggestion
The compiler also ensures that the input code is valid Swift;
```

> for example, that the input to macros is syntactically valid

Reviewer: ```suggestion
for example, that the input to a macro is syntactically valid
```

> and that the types of values match.

Reviewer: value types?

For example, in the code above,

> it checks that the argument to `#fourCharacterCode` is a string.

Reviewer: ```suggestion
the compiler checks that the argument to `#fourCharacterCode` is a string.
```

<!-- XXX trying to avoid jargon "type checking" here -->

The compiler finds the places in the code where you call a macro,
and loads the external binary that implements those macros.
For each macro call,
the compiler passes part of the AST to that macro's implementation.
<!-- behind the scenes, the AST is serialized thru JSON -->
Here's a representation of that partial AST:

[XXX placeholder art]

- Macro
    - Identifier: `fourCharacterCode`
    - Arguments
        - String literal `"ABCD"`

The implementation of a macro
operates only on the code that contains the macro.
In this example,
the implementation of the `#fourCharacterCode` macro
reads this partial AST that contains only the macro call.
A macro can't depend on code outside of the place where it's used.

The implementation of `#fourCharacterCode`
generates a new AST containing the expanded code.
Here's what that code returns to the compiler:

[XXX placeholder art]

- Integer literal: `1145258561`

> When the compiler gets this expansion back,

Reviewer: ```suggestion
When the compiler receives this expansion,
```

it replaces the AST element that contains the macro call
with the element that contains the macro's expansion.
After macro expansion,
the compiler checks again to ensure
the program is still syntactically valid Swift

> and the all the types are correct.

Reviewer: ```suggestion
and that all of the types are correct.
```

That produces a final AST that can be compiled as usual:

[XXX placeholder art]

- Constant declaration
    - Identifier: `magicNumber`
    - Value
        - Integer literal: `1145258561`

This AST corresponds to Swift code like this:

```
let magicNumber = 1145258561
```

In this example, the input source code has only one macro,
but a real program could have several instances of the same macro
and several calls to different macros.
The compiler expands macros one at a time.

If one macro appears inside another,

> the outer macro is expanded first ---

Reviewer: ```suggestion
the outer macro expands first ---
```

this lets the outer macro modify the inner macro before it's expanded.

<!--

OUTLINE bits for the future

- Macros can be nested.
  Nested macros are expanded from the outside in.
  XXX TR: Is there any limit to nesting?
  XXX TR: Is it valid to nest like this -- if so, anything to note about it?

  ```
  let something = #someMacro {
      struct A { }
      @someMacro struct B { }
  }
  ```

- Macro recursion is limited.
  One macro can call another,
  but a given macro can't directly or indirectly call itself.
  The result of macro expansion can include other macros,
  but it can't include a macro that uses this macro in its expansion
  or declare a new macro.
  (XXX likely need to iterate with Doug here)

-->

## Implementing a Macro

> To implement a macro, you make two parts:

Reviewer: Instead of "parts," can we use "components"?

A type that performs the macro expansion,
and a library that declares the macro to expose it as API.
These parts are built separately from code that uses the macro,
even if you're developing the macro and its clients together,
because the macro implementation runs
as part of building the macro's clients.

To create a new macro using Swift Package Manager,
run `swift package init --type macro` ---
this creates several files,
including a template for a macro implementation and declaration.
To add macros to an existing project,
add a target for the macro implementation
and a target for the macro library.
For example,
you can add something like the following to your `Package.swift` file,
changing the names to match your project:

```swift
targets: [

>     // Macro implementation that performs the source transformations

Reviewer: ```suggestion
    // Macro implementation that performs the source transformations.
```

    .macro(
        name: "MyProjectMacros",
        dependencies: [
            .product(name: "SwiftSyntaxMacros", package: "swift-syntax"),
            .product(name: "SwiftCompilerPlugin", package: "swift-syntax")
        ]
    ),

>     // Library that exposes a macro as part of its API

Reviewer: ```suggestion
    // Library that exposes a macro as part of its API.
```

    for character in characters {
        result = result << 8
        guard let asciiValue = character.asciiValue else { return nil }
        result += UInt32(asciiValue)
    }
    return result.bigEndian
}
```

This is a freestanding macro that produces an expression,
so the `FourCharacterCode` type that implements the macro
conforms to the `ExpressionMacro` protocol.
The `ExpressionMacro` protocol has one requirement,
a `expansion(of:in:)` method that expands the AST.
For the list of macro roles and their corresponding SwiftSystem protocols,
see <doc:Attributes#attached> and <doc:Attributes:freestanding>
in <doc:Attributes>

To expand the `#fourCharacterCode` macro,
Swift sends the AST for the code that uses this macro
to the library that contains the macro implementation.
Inside the library, Swift calls `FourCharacterCode.expansion(of:in:)`,
passing in the AST and the context as arguments to the method.
The implementation of `expansion(of:in:)`
finds the string that was passed as an argument to `#fourCharacterCode`
and calculates the corresponding integer literal value.

In the example above,
the first `guard` block extracts the string literal from the AST,

> and the second `gaurd` block

Reviewer: ```suggestion
and the second `guard` block
```

calls the private `FourCharacterCode(for:)` function.

> Both of these blocks of code throw an error if the macro is use incorrectly ---

Reviewer: ```suggestion
Both of these blocks throw an error if the macro is used incorrectly ---
```

the error message becomes a compiler error
at the malformed call site.
For example,
if you try to call the macro as `#fourCharacterCode("AB" + "CD")`
the compiler shows the error "Need a static string".
<!-- XXX generate good diagnostics when your macro is used wrong -->

The `expansion(of:in:)` method returns an instance of `ExprSyntax`,
a type from SwiftSyntax that represents an expression in an AST.
Because this type conforms to the `StringLiteralConvertible` protocol,
the macro implementation uses a string literal
as a lightweight syntax to create its result.
All of the SwiftSyntax types that you return from a macro implementation
conform to `StringLiteralConvertible`,

> so you can use this approach when implementing any kind  of macro.

Reviewer: ```suggestion
so you can use this approach when implementing any kind of macro.
```

  or `SyntaxToken`
  or `ExprSyntax`.
  (Need to give folks some general ideas,
  and enough guidance so they can sort through
  all the various `SwiftSyntax` node types and find the right one.)

- Attached macros follow the same general model as expression macros,
  but with more moving parts.

- Pick the subprotocol of `AttachedMacro` to conform to,
  depending on which kind of attached macro you're making.
  [This is probably a table]

  + `AccessorMacro` goes with `@attached(accessor)`
  + `ConformanceMacro` goes with `@attached(conformance)`
    [XXX missing from the list under Declaring a Macro]
  + `MemberMacro` goes with `@attached(member)`
  + `PeerMacro` goes with `@attached(peer)`
  + `MemberAttributeMacro` goes with `@member(memberAttribute)`

- Code example of conforming to `MemberMacro`.

  ```
  static func expansion<
    Declaration: DeclGroupSyntax,
    Context: MacroExpansionContext
  >(
    of node: AttributeSyntax,
    providingMembersOf declaration: Declaration,
    in context: Context
  ) throws -> [DeclSyntax]
  ```

- Adding a new member by making an instance of `Declaration`,
  and returning it as part of the `[DeclSyntax]` list.

-->

## Debugging Macros

> Macros are well suited to development using tests:

Reviewer: ```suggestion
Macros are well suited to testing:
```

> they transform one AST into another AST,

Reviewer: ```suggestion
they transform one AST into another AST
```

without depending on any external state,

> and without causing any any side effects.

Reviewer: Such as?

<!-- XXX Frame differently; this is motivation, not an example -->
For example, consider the following code
that doesn't use macros:

```swift
struct SundaeToppings: OptionSet {
    let rawValue: Int
    static let nuts = SundaeToppings(rawValue: 1 << 0)
    static let cherry = SundaeToppings(rawValue: 1 << 1)
    static let fudge = SundaeToppings(rawValue: 1 << 2)
}
```

In this code,
each of the options in the `SundaeToppings` option set
includes a call to the initializer,
which is repetitive and manual.
It would be easy to make a mistake when adding a new option,
like typing the wrong number at the end of the line.

Here's a version of this code that uses a macro instead:

```swift
@OptionSet<Int>
struct SundaeToppings {
    private enum Options: Int {
        case nuts
        case cherry
        case fudge
    }
}
```

This version of `SundaeToppings`
calls the [`@OptionSet`][] macro from the Swift standard library.
The macro reads the list of cases in the private enumeration,
generates the list of constants for each option,
and adds a conformance to the [`OptionSet`][] protocol.

[`@OptionSet`]: https://developer.apple.com/documentation/swift/optionset-swift.macro

> [`OptionSet`]: https://developer.apple.com/documentation/swift/optionset-swift.protocol

Author: Hmm… I see the `OptionSet` macro in source on the Swift 5.9 release branch:

https://github.com/apple/swift/blob/release/5.9-20230510/stdlib/public/core/Macros.swift#L93

```
public macro OptionSet<RawType>() = #externalMacro(module: "SwiftMacros", type: "OptionSetMacro")
```

Git blame shows it was added by [this commit](https://github.com/apple/swift/commit/0e9c3eff8a534218e013a355449bf23b9260176d) which is from [PR 6409](https://github.com/apple/swift/pull/64091).

The links should be correct, yes.  DocC requires some a disambiguation suffix in cases like this, where there's a macro and a protocol with the same name, which I had to ask some folks about.  I'd expect the existing URL to continue to work, but I'll confirm.
